plutus-core/plutus-ir/src/PlutusIR/Analysis/RetainedSize.hs

```diff
@@ -0,0 +1,215 @@
+bindingSizes term                 = term ^. termSubterms . to bindingSizes
+
+-- | Same as 'bindingSizes' but is wrapped in a newtype and has a bogus entry for the root.
+toDirectionRetentionMap
+    :: (HasUnique tyname TypeUnique, HasUnique name TermUnique)
+    => Term tyname name uni fun ann -> DirectionRetentionMap
+toDirectionRetentionMap term =
+    DirectionRetentionMap . IntMap.insert rootInt rootSize . unUniqueMap $ bindingSizes term where
+        -- See Note [Handling the root].
+        rootSize = Size (- 10 ^ (10::Int))
+
+-- | Check if a 'Node' appears in 'DirectionRetentionMap'.
+hasSizeIn :: DirectionRetentionMap -> Node -> Bool
+hasSizeIn _             Root                                   = True
+hasSizeIn (DirectionRetentionMap ss) (Variable (PLC.Unique i)) = i `IntMap.member` ss
+
+-- | Compute the retention map of a term.
+termRetentionMap
+    :: (HasUnique tyname TypeUnique, HasUnique name TermUnique, ToBuiltinMeaning uni fun)
+    => Term tyname name uni fun ann -> IntMap Size
+termRetentionMap term = depsRetentionMap sizeInfo deps where
+    sizeInfo = toDirectionRetentionMap term
+    deps = C.induce (hasSizeIn sizeInfo) $ fst $ runTermDeps term
+
+-- | Apply a function to the annotation of each part of every 'Binding' in a term.
+reannotateBindings
+    :: (HasUnique name TermUnique, HasUnique tyname TypeUnique)
+    => (Unique -> ann -> ann)
+    -> Term tyname name uni fun ann
+    -> Term tyname name uni fun ann
+reannotateBindings f = goLet where
+    -- We don't need these helper functions anywhere else, so we make them into local definitions.
+    goVarDecl (VarDecl ann name ty) = VarDecl (f (name ^. theUnique) ann) name ty
+    goTyVarDecl (TyVarDecl ann tyname kind) = TyVarDecl (f (tyname ^. theUnique) ann) tyname kind
+    goDatatype (Datatype ann dataTyDecl paramTyDecls matchName constrDecls) =
+        Datatype
+            (f (matchName ^. theUnique) ann)
+            (goTyVarDecl dataTyDecl)
+            (goTyVarDecl <$> paramTyDecls)
+            matchName
+            (goVarDecl <$> constrDecls)
+
+    -- Note that @goBind@ and @goLet@ are mutually recursive.
+    goBind (TermBind ann str var term) = TermBind ann str (goVarDecl var) $ goLet term
+    goBind (TypeBind ann tyVar ty)     = TypeBind ann (goTyVarDecl tyVar) ty
+    goBind (DatatypeBind ann datatype) = DatatypeBind ann $ goDatatype datatype
+
+    goLet (Let ann recy binds term) = Let ann recy (goBind <$> binds) $ goLet term
+    goLet term                      = term & termSubterms %~ goLet
+
+-- See Note [Retained size analysis]
+-- | Annotate each part of every 'Binding' in a term with the size that it retains.
+annotateWithRetainedSize
+    :: (HasUnique name TermUnique, HasUnique tyname TypeUnique, ToBuiltinMeaning uni fun)
+    => Term tyname name uni fun ann
+    -> Term tyname name uni fun RetainedSize
+annotateWithRetainedSize term = reannotateBindings (upd . unUnique) $ NotARetainer <$ term where
```

Comment:
Maybe worth a comment about the annotation processing:
- `reannotateBindings` only processes annotations "associated with" a unique, so it can't change the type
- Therefore we need to set all the bindings to an appropriate type beforehand, hence this step.

This made me wonder about a slightly different approach:
- Write a more custom processing step that replaces *all* the annotations with a `Maybe Unique` if they're associated with an annotation.
- Replace those with retained sizes wholesale using `fmap`.

Having written that, I'm not sure it's any better: you still need the custom traversal.  Third idea:
- Give `reannotateBinding` another `ann -> ann` argument which gets called on annotations not associated with a unique. Now you can make it change the type of annotations, and do this in one step.

Probably none of this is worth it, but I've written this comment now so :shrug:

Reply:
> Write a more custom processing step that replaces all the annotations with a `Maybe Unique` if they're associated with an annotation.

Given that we do a similar thing during scoping analysis, I believe it's exactly the right thing to do. But I'm too lazy to perform this refactoring, I think it would be fine as a separate PR.

> Give reannotateBinding another ann -> ann argument which gets called on annotations not associated with a unique. Now you can make it change the type of annotations, and do this in one step.

Then I'd need to match on every constructor and manually change every annotation instead of just using the nice `    goLet term = term & termSubterms %~ goLet` stuff.